src/Core/Managed/Shared/Extensibility/Metric.cs

```diff
@@ -0,0 +1,171 @@
+
+            this.aggregatorId = Metric.GetAggregatorId(name, dimensions);
+            this.hashCode = this.aggregatorId.GetHashCode();
+        }
+
+        /// <summary>
+        /// Gets metric name.
+        /// </summary>
+        public string Name { get; private set; }
+
+        /// <summary>
+        /// Gets a set of metric dimensions and their values.
+        /// </summary>
+        public IDictionary<string, string> Dimensions { get; private set; }
+
+        /// <summary>
+        /// Adds a value to the time series.
+        /// </summary>
+        /// <param name="value">Metric value.</param>
+        public void Track(double value)
+        {
+            SimpleMetricStatisticsAggregator aggregator = this.manager.GetStatisticsAggregator(this);
+            aggregator.Track(value);
+
+            this.ForwardToProcessors(value);
+        }
+
+        /// <summary>
+        /// Returns the hash code for this object.
+        /// </summary>
+        /// <returns>A 32-bit signed integer hash code.</returns>
+        public override int GetHashCode()
+        {
+            return this.hashCode;
+        }
+
+        /// <summary>
+        /// Determines whether the specified object is equal to the current object.
+        /// </summary>
+        /// <param name="other">The object to compare with the current object. </param>
+        /// <returns>True if the specified object is equal to the current object; otherwise, false.</returns>
+        public bool Equals(Metric other)
+        {
+            if (other == null)
+            {
+                return false;
+            }
+
+            return string.Compare(this.aggregatorId, other.aggregatorId, StringComparison.Ordinal) == 0;
+        }
+
+        /// <summary>
+        /// Determines whether the specified object is equal to the current object.
+        /// </summary>
+        /// <param name="obj">The object to compare with the current object. </param>
+        /// <returns>True if the specified object is equal to the current object; otherwise, false.</returns>
+        public override bool Equals(object obj)
+        {
+            if (ReferenceEquals(null, obj))
+            {
+                return false;
+            }
+
+            if (ReferenceEquals(this, obj))
+            {
+                return true;
+            }
+
+            return this.Equals(obj as Metric);
+        }
+
+        /// <summary>
+        /// Generates id of the aggregator serving time series specified in the parameters.
+        /// </summary>
+        /// <param name="name">Metric name.</param>
+        /// <param name="dimensions">Optional metric dimensions.</param>
+        /// <returns>Aggregator id that can be used to get aggregator.</returns>
+        private static string GetAggregatorId(string name, IDictionary<string, string> dimensions = null)
+        {
+            StringBuilder aggregatorIdBuilder = new StringBuilder(name ?? string.Empty);
+
```

Comment:
Consider a different string constant than "" here. This might help debugging. Can name actually be empty? If yes, we will confuse it with a null name. Maybe "MetricNameIsNull" or somthing?